### PR TITLE
Linearizability raft before leader send

### PR DESCRIPTION
**Pull request by serathius**
Part of https://github.com/etcd-io/etcd/issues/14726
cc @ahrtr

### CODE REVIEW

**Comment by ahrtr** [comment]
Please resolve the pipeline failures.

**Comment by serathius** [comment]
@ahrtr done

File: tests/framework/e2e/cluster.go [hunk 1]
```diff
@@ -44,6 +44,13 @@ const (
 	ClientTLSAndNonTLS
 )
 
+type ClientConfig struct {
+	ConnectionType            ClientConnType
+	CertificateAuthority      bool
```
Comment (ahrtr) [inline tests/framework/e2e/cluster.go:49]:
```suggestion
	CertAuthEnabled      bool
```

File: tests/framework/e2e/cluster.go [hunk 2]
```diff
@@ -44,6 +44,13 @@ const (
...
+	AutogenerateCertificates  bool
+	CertificateRevocationList bool
```
Comment (ahrtr) [inline tests/framework/e2e/cluster.go:51]:
```suggestion
	CertificateRevoked bool
```

File: tests/framework/e2e/cluster.go [hunk 3]
```diff
@@ -142,13 +149,11 @@ type EtcdProcessClusterConfig struct {
...
-	ClientCertAuthEnabled bool
-	IsPeerTLS             bool
-	IsPeerAutoTLS         bool
-	IsClientAutoTLS       bool
-	IsClientCRL           bool
-	NoCN                  bool
+	Client        ClientConfig
+	IsPeerTLS     bool
+	IsPeerAutoTLS bool
+	IsClientCRL   bool
```
Comment (ahrtr) [inline tests/framework/e2e/cluster.go:155]: Why do you still keep this field? Isn't it included in ClientConfig? 
```suggestion
	IsClientCRL   bool
```

File: tests/framework/e2e/cluster.go [hunk 4]
```diff
@@ -142,13 +149,11 @@ type EtcdProcessClusterConfig struct {
...
-	ClientTLS             ClientConnType
-	ClientCertAuthEnabled bool
-	IsPeerTLS             bool
-	IsPeerAutoTLS         bool
-	IsClientAutoTLS       bool
-	IsClientCRL           bool
-	NoCN                  bool
+	Client        ClientConfig
+	IsPeerTLS     bool
+	IsPeerAutoTLS bool
```
Comment (ahrtr) [inline tests/framework/e2e/cluster.go:154]: Probably we can add a similar `PeerConfig`?

**Comment by serathius** [comment]
Please take another look @ahrtr

File: tests/linearizability/failpoints.go [hunk 5]
```diff
@@ -32,24 +33,25 @@ import (
...
+	BackendAfterWritebackBufPanic            Failpoint = goFailpoint{"backend/afterWritebackBuf", "panic", nil, AnyMember}
+	CompactBeforeCommitScheduledCompactPanic Failpoint = goFailpoint{"mvcc/compactBeforeCommitScheduledCompact", "panic", triggerCompact, AnyMember}
+	CompactAfterCommitScheduledCompactPanic  Failpoint = goFailpoint{"mvcc/compactAfterCommitScheduledCompact", "panic", triggerCompact, AnyMember}
+	CompactBeforeSetFinishedCompactPanic     Failpoint = goFailpoint{"mvcc/compactBeforeSetFinishedCompact", "panic", triggerCompact, AnyMember}
+	CompactAfterSetFinishedCompactPanic      Failpoint = goFailpoint{"mvcc/compactAfterSetFinishedCompact", "panic", triggerCompact, AnyMember}
+	CompactBeforeCommitBatchPanic            Failpoint = goFailpoint{"mvcc/compactBeforeCommitBatch", "panic", triggerCompact, AnyMember}
+	CompactAfterCommitBatchPanic             Failpoint = goFailpoint{"mvcc/compactAfterCommitBatch", "panic", triggerCompact, AnyMember}
+	RaftBeforeLeaderSendPanic                Failpoint = goFailpoint{"etcdserver/raftBeforeLeaderSend", "panic", nil, Leader}
```
Comment (ahrtr) [inline tests/linearizability/failpoints.go:54]: There are too many duplicated parameters, e.g. `nil`, `AnyMember`. We should only provide a value when it isn't `nil` or `AnyMember`. Let's consider to refactor it later in separate PRs.

**Review by ahrtr — APPROVED** [review]
Overall looks good to me.

File: tests/linearizability/failpoints.go (hunk 5 above)
Comment (serathius) [reply tests/linearizability/failpoints.go:54]: Agree, I'm thinking about how to combine this with allowing non-panic failpoints

**Pull request merged**